rustls-mbedcrypto-provider/src/hash.rs

```diff
@@ -0,0 +1,154 @@
+#[cfg(feature = "logging")]
+use crate::log::error;
+use alloc::boxed::Box;
+use alloc::sync::Arc;
+use alloc::vec;
+use alloc::vec::Vec;
+use rustls::crypto::hash::{self, HashAlgorithm};
+use std::sync::Mutex;
+
+pub(crate) static SHA256: Hash = Hash(&MBED_SHA_256);
+pub(crate) static SHA384: Hash = Hash(&MBED_SHA_384);
+
+pub(crate) struct Hash(&'static Algorithm);
+
+/// A digest algorithm.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct Algorithm {
+    pub(crate) hash_algorithm: HashAlgorithm,
+    pub(crate) hash_type: mbedtls::hash::Type,
+    pub(crate) output_len: usize,
+}
+
+/// SHA-256 as specified in [FIPS 180-4].
+///
+/// [FIPS 180-4]: http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf
+pub(crate) static MBED_SHA_256: Algorithm = Algorithm {
+    hash_algorithm: HashAlgorithm::SHA256,
+    hash_type: mbedtls::hash::Type::Sha256,
+    output_len: 256 / 8,
+};
+
+/// SHA-384 as specified in [FIPS 180-4].
+///
+/// [FIPS 180-4]: http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf
+pub(crate) static MBED_SHA_384: Algorithm = Algorithm {
+    hash_algorithm: HashAlgorithm::SHA384,
+    hash_type: mbedtls::hash::Type::Sha384,
+    output_len: 384 / 8,
+};
+
+impl hash::Hash for Hash {
+    fn start(&self) -> Box<dyn hash::Context> {
+        Box::new(HashContext(MbedHashContext::new(self.0)))
+    }
+
+    fn hash(&self, data: &[u8]) -> hash::Output {
+        hash::Output::new(&hash(self.0, data))
+    }
+
+    fn algorithm(&self) -> HashAlgorithm {
+        self.0.hash_algorithm
+    }
+
+    fn output_len(&self) -> usize {
+        self.0.output_len
+    }
+}
+
+struct HashContext(MbedHashContext);
+
+impl hash::Context for HashContext {
+    fn fork_finish(&self) -> hash::Output {
+        hash::Output::new(&self.0.clone().finalize())
+    }
+
+    fn fork(&self) -> Box<dyn hash::Context> {
+        Box::new(Self(self.0.clone()))
+    }
+
+    fn finish(self: Box<Self>) -> hash::Output {
+        hash::Output::new(&self.0.finalize())
+    }
+
+    fn update(&mut self, data: &[u8]) {
+        self.0.update(data)
+    }
+}
+
+pub(crate) struct MbedHashContext {
+    pub(crate) state: Arc<Mutex<mbedtls::hash::Md>>,
```

Comment:
Nice catch.
Arc is indeed removeable, will remove it.

> BTW, it's not clear to me why `Md` could not be `Sync`, especially since its methods take `&mut self`, and it is `Send` already. But that's a different story I suppose.

This is right, but need to double check all methods that take `&self` are thread safe on C side and the rust code does not use any pointer conversion. 

I will do this potential improvement in another PR.